src/WebJobs.Script.WebHost/Filters/AuthorizationLevelAttribute.cs

```diff
@@ -44,32 +58,35 @@ public AuthorizationLevelAttribute(AuthorizationLevel level)
                 // as a request property
                 var secretManager = actionContext.ControllerContext.Configuration.DependencyResolver.GetService<ISecretManager>();
 
-                requestAuthorizationLevel = await GetAuthorizationLevelAsync(request, secretManager, EvaluateKeyMatch);
-                request.SetAuthorizationLevel(requestAuthorizationLevel);
+                var result = await GetAuthorizationResultAsync(request, secretManager, EvaluateKeyMatch, KeyName);
+                requestAuthorizationLevel = result.AuthorizationLevel;
+                request.SetAuthorizationLevel(result.AuthorizationLevel);
+                request.SetProperty(ScriptConstants.AzureFunctionsHttpRequestKeyNameKey, result.KeyName);
             }
 
             if (request.IsAuthDisabled() ||
                 SkipAuthorization(actionContext) ||
                 Level == AuthorizationLevel.Anonymous)
             {
+                // no authorization required
                 return;
             }
 
-            if (requestAuthorizationLevel < Level)
+            if (!request.HasAuthorizationLevel(Level))
             {
                 actionContext.Response = new HttpResponseMessage(HttpStatusCode.Unauthorized);
             }
         }
 
-        protected virtual bool EvaluateKeyMatch(IDictionary<string, string> secrets, string keyValue) => HasMatchingKey(secrets, keyValue);
+        protected virtual string EvaluateKeyMatch(IDictionary<string, string> secrets, string keyName, string keyValue) => GetKeyMatchOrNull(secrets, keyName, keyValue);
 
-        internal static Task<AuthorizationLevel> GetAuthorizationLevelAsync(HttpRequestMessage request, ISecretManager secretManager, string functionName = null)
+        internal static Task<KeyAuthorizationResult> GetAuthorizationResultAsync(HttpRequestMessage request, ISecretManager secretManager, string keyName = null, string functionName = null)
```

Comment:
All this refactoring to return a structured result containing the matched key ID supports the changes in this PR, but will also be needed for the EasyAuth updates I'll be doing soon (where we want to be able to flow the matched KeyId to the user code).